cmd/httpapi/main.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	tlsConfig, CACertChain, err := mtlsConfig.InitCertificates()
+	if err != nil {
+		logger.Errorf("Cannot retrieve any MTLS configuration: %w", err)
+		os.Exit(1)
+	}
+
+	// @TODO check here what to do with leftovers or if a new one is need to be created
+	err = mtlsConfig.CreateRegistrationClientCerts()
+	if err != nil {
+		logger.Errorf("Cannot create registration client certificate: %w", err)
+		os.Exit(1)
+	}
+
+	opts := x509.VerifyOptions{
+		Roots:         tlsConfig.ClientCAs,
+		Intermediates: x509.NewCertPool(),
+	}
+
+	k8sClient := k8sclient.NewK8sClient(c)
+	edgeDeviceSignedRequestRepository := edgedevicesignedrequest.NewEdgedeviceSignedRequestRepository(c)
+	edgeDeviceRepository := edgedevice.NewEdgeDeviceRepository(c)
+	edgeWorkloadRepository := edgeworkload.NewEdgeWorkloadRepository(c)
+	edgeDeviceSetRepository := edgedeviceset.NewEdgeDeviceSetRepository(c)
+	claimer := storage.NewClaimer(c)
+
+	metricsObj := metrics.New()
+
+	corev1Client, err := corev1client.NewForConfig(clientConfig)
+	if err != nil {
+		panic(err)
+	}
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&v1.EventSinkImpl{Interface: corev1Client.Events("")})
+	defer func() {
+		broadcaster.Shutdown()
+	}()
+
+	yggdrasilAPIHandler := yggdrasil.NewYggdrasilHandler(
+		edgeDeviceSignedRequestRepository,
+		edgeDeviceRepository,
+		edgeWorkloadRepository,
+		edgeDeviceSetRepository,
+		claimer,
+		k8sClient,
+		initialDeviceNamespace,
+		broadcaster.NewRecorder(scheme, corev1.EventSource{Component: "flotta-edge-api"}),
+		registryAuth,
+		metricsObj,
+		devicemetrics.NewAllowListGenerator(k8sClient),
+		configmaps.NewConfigMap(k8sClient),
+		mtlsConfig,
+		logger,
+	)
+
+	var api *operations.FlottaManagementAPI
+	var handler http.Handler
+
+	APIConfig := restapi.Config{
+		YggdrasilAPI: yggdrasilAPIHandler,
+		InnerMiddleware: func(h http.Handler) http.Handler {
+			// This is needed for one reason. Registration endpoint can be
+			// triggered with a certificate signed by the CA, but can be expired
+			// The main reason to allow expired certificates in this endpoint, it's
+			// to renew client certificates, and because some devices can be
+			// disconnected for days and does not have the option to renew it.
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if r.TLS == nil {
+					w.WriteHeader(http.StatusBadRequest)
+					return
+				}
+
+				authType := yggdrasilAPIHandler.GetAuthType(r, api)
+				if ok, err := mtls.VerifyRequest(r, authType, opts, CACertChain, yggdrasil.AuthzKey, logger); !ok {
+					metricsObj.IncEdgeDeviceFailedAuthenticationCounter()
+					logger.Info("cannot verify request:", "authType", authType, "method", r.Method, "url", r.URL, "err", err)
+					w.WriteHeader(http.StatusUnauthorized)
+					return
+				}
+				h.ServeHTTP(w, r)
+			})
+		},
+	}
+	handler, api, err = restapi.HandlerAPI(APIConfig)
+	if err != nil {
+		logger.Errorf("cannot start http server: %w", err)
```

Comment:
Why are you keep going if there is an error? Shouldn't be a `os.Exit` here?

Reply:
Good point.